### PR TITLE
Correct suggestion for println

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-lang/rust/issues/51585
r? @estebank

### CODE REVIEW

File: src/test/ui/macros/bad_hello.stderr [hunk 1]
```diff
@@ -2,7 +2,7 @@ error: expected a literal
   --> $DIR/bad_hello.rs:12:14
    |
 LL |     println!(3 + 4); //~ ERROR expected a literal
-   |              ^^^^^
+   |              ^^^^^ help: consider changing this to: `"{}", 3 + 4`
```
Comment (varkor) [inline src/test/ui/macros/bad_hello.stderr]: I think this suggestion as-is could be a little confusing: selecting the entire expression and suggesting it be replaced with `println!("{}", 3 + 4)` would give more helpful context.

File: src/libsyntax_ext/concat.rs [hunk 2]
```diff
@@ -53,7 +54,13 @@ pub fn expand_syntax_ext(cx: &mut base::ExtCtxt,
                 }
             }
             _ => {
-                cx.span_err(e.span, "expected a literal");
+                let mut err = cx.struct_span_err(e.span, "expected a literal");
+                err.span_suggestion(
+                    e.span,
+                    "consider changing this to",
+                    format!("\"{{}}\", {}", pprust::expr_to_string(&e))
```
Comment (estebank) [inline src/libsyntax_ext/concat.rs]: Could you use span to snippet here and only use expression to string only if that call fails? We want to preserve the code as is, like for example whitespace.

File: src/libsyntax_ext/concat.rs [hunk 3]
```diff
@@ -53,7 +54,13 @@ pub fn expand_syntax_ext(cx: &mut base::ExtCtxt,
...
+                );
+                err.emit();
```
Comment (estebank) [inline src/libsyntax_ext/concat.rs]: Could you add a note explaining that a format string is always needed?

**Comment by TimNN** [comment]
Ping form triage, @csmoe: It looks like there are some comments on your PR, do you think you'll be able to address those?

**Comment by csmoe** [comment]
@TimNN testing locally, more commits later.

**Comment by TimNN** [comment]
Ping from triage, @estebank ! This PR requires your review.

File: src/libsyntax_ext/concat.rs [hunk 4]
```diff
@@ -53,7 +53,12 @@ pub fn expand_syntax_ext(cx: &mut base::ExtCtxt,
                 }
             }
             _ => {
-                cx.span_err(e.span, "expected a literal");
+                let mut err = cx.struct_span_err(e.span, "expected a literal");
+                err.span_help(
```
Comment (estebank) [inline src/libsyntax_ext/concat.rs]: This should be

```
if let Some(snippet) = cx.codemap().span_to_snippet(span) { 
    err.span_suggestion(
        e.span,
        "you might be missing a string literal to format with",
        format!("\"{{}}\", {}", snippet);
    );
}
```

Which will give you the following output:

```

error: expected a literal
  --> $DIR/bad_hello.rs:12:14
   |
LL |     println!(3 + 4); //~ ERROR expected a literal
   |              ^^^^^
   |
help: you might be missing a string literal to format with
  --> $DIR/bad_hello.rs:12:14
   |
LL |     println!("{}", 3 + 4); //~ ERROR expected a literal
   |              ^^^^^^^^^^^
```

**Comment by estebank** [comment]
Oh, sorry, it is a `Result`, not an `Option` :-/ (I think it changed "recently")

**Comment by csmoe** [comment]
@estabank
I tried some same solution as you suggested before and the test failed locally as above, then I gave up that, I thought there may be no way to address your advice since we are working with `concat!`:
concat_span:
https://github.com/rust-lang/rust/blob/4700e1188f66fdb4086b7593416b678b8fe935f4/src/libsyntax_ext/concat.rs#L20-L23
expr_span:
https://github.com/rust-lang/rust/blob/4700e1188f66fdb4086b7593416b678b8fe935f4/src/libsyntax_ext/concat.rs#L51
so the span of `println!` which you and varkor are asking for span_suggestion cannot be derived here(if else I need your tips).

**Pull request merged**